crates/ruff_python_formatter/src/expression/string/docstring.rs

```diff
@@ -702,19 +784,336 @@ impl<'src> CodeExampleDoctest<'src> {
         //
         // [1]: https://github.com/python/cpython/blob/0ff6368519ed7542ad8b443de01108690102420a/Lib/doctest.py#L733
         if self.ps1_indent != ps2_indent {
-            return false;
+            queue.push_back(self.into_format_action());
+            return None;
         }
         // PS2 prompts must be followed by an ASCII space character unless
         // it's an otherwise empty line[1].
         //
         // [1]: https://github.com/python/cpython/blob/0ff6368519ed7542ad8b443de01108690102420a/Lib/doctest.py#L809-L812
         let code = match ps2_after.strip_prefix(' ') {
             None if ps2_after.is_empty() => "",
-            None => return false,
+            None => {
+                queue.push_back(self.into_format_action());
+                return None;
+            }
             Some(code) => code,
         };
         self.lines.push(CodeExampleLine { original, code });
-        true
+        queue.push_back(CodeExampleAddAction::Kept);
+        Some(self)
+    }
+
+    /// Consume this doctest and turn it into a formatting action.
+    fn into_format_action(self) -> CodeExampleAddAction<'src> {
+        CodeExampleAddAction::Format {
+            kind: CodeExampleKind::Doctest(self),
+        }
+    }
+}
+
+/// State corresponding to a single reStructuredText literal block or
+/// code-block directive.
+///
+/// While a literal block and code-block directive are technically two
+/// different reStructuredText constructs, we use one type to represent
+/// both because they are exceptionally similar. Basically, they are
+/// the same with two main differences:
+///
+/// 1. Literal blocks are began with a line that ends with `::`. Code block
+/// directives are began with a line like `.. code-block:: python`.
+/// 2. Code block directives permit a list of options as a "field list"
+/// immediately after the opening line. Literal blocks have no options.
+///
+/// Otherwise, everything else, including the indentation structure, is the
+/// same.
+#[derive(Debug)]
+struct CodeExampleRst<'src> {
+    /// The lines that have been seen so far that make up the block.
+    lines: Vec<CodeExampleLine<'src>>,
+    /// The indent of the line "opening" this block. It can either be the
```

Comment:
Nit: I believe we use newlines between fields for most of our code base but I'm okay leaving it as is, considering that this isn't enforced by a formatter. Is it your preference to omit new lines between fields (I also noticed that you omit newlines between control flow blocks).

Reply:
I do indeed omit empty lines between fields in type definitions. I think I've just perceived that to be the common style, but looking a bit, the standard library seems to use a wild mix haha.

As for control flow blocks, I have less of a consistent style there. I tend to use empty lines when logical groupings make sense. But that's just somewhat of a judgment call. Looking a bit, probably some empty lines would help in places. But no, my preference wouldn't be to always insert an empty line.

With that said, I like consistent styling, so I'll add some line breaks here and try to remember to do so in the future. But yeah, without the formatter doing this for us, it's bound to get out-of-whack.